src/lib/components/Graph/dagre.js

```diff
@@ -15,183 +14,253 @@ const config = {
+// Dagre-d3.
+//
+// The `drawGraph` function is exported, but the `Dagre` class is where
+// most of the logic lives. The class is instantiated with the nodes, links,
+// previous nodes, previous links, a callback that should be called when a node
+// is clicked, and some d3 elements.
+//
+// The `draw` function works like a state machine and executes all the logic
+// necessary for determining whether the graph should be redrawn and how to draw
+// it. For example, if several nodes have been added, the entire graph is
+// redrawn. But if the only change is the active node, the graph is not redrawn.
+// Instead, the CSS classes of the active node and previously active node are
+// changed.
+//
+// See the comments in the `draw` function for more information.
+class Dagre {
+  constructor(selectedSvg, selectedG, nodes, links, onClick, prevNodes, prevLinks) {
+    this.selectedSvg = selectedSvg;
+    this.selectedG = selectedG;
+    this.nodes = nodes;
+    this.links = links;
+    this.onClick = onClick;
+    this.prevNodes = prevNodes;
+    this.prevLinks = prevLinks;
+
+    this.state = 'initial';
+
+    // Instance variables that can be set or unset as part of a state transition
+    this.active = null;
+    this.previousActive = null;
+
+    this.draw = this.draw.bind(this);
+    this.checkChanges = this.checkChanges.bind(this);
+    this.redraw = this.redraw.bind(this);
+    this.checkLinks = this.checkLinks.bind(this);
+    this.checkNodes = this.checkNodes.bind(this);
+    this.drawActive = this.drawActive.bind(this);
   }
 
-  if (nodes.length !== prevNodes.length) {
-    return false;
-  }
+  draw() {
+    if (this.state === 'complete') {
+      return;
+    }
 
-  if (links.length !== prevLinks.length) {
-    return false;
+    const transitions = {
+      // Checks for surface-level changes (e.g. number of nodes)
+      // -> links  : if there are no changes
+      // -> redraw : if there are changes
+      initial: this.checkChanges,
+
+      // Checks for any changes between `links` and `prevLinks`
+      // -> nodes  : if there are no changes
+      // -> redraw : if there are changes
+      links: this.checkLinks,
+
+      // Checks for any changes between `nodes` and `prevNodes`
+      // Also checks whether the active node has changed
+      // (Sets `this.active` and `this.previousActive`)
+      // -> complete : if there are no changes
+      // -> active   : if the only change is the active node
+      // -> redraw   : if there are any other changes
+      nodes: this.checkNodes,
+
+      // Adds the `active` class to the active node in the graph
+      // Removes the `active` class from the previously active node
+      // Makes no other change to the graph
+      // -> complete
+      active: this.drawActive,
+
+      // Draws the graph
+      // If the graph has already been drawn, this function redraws it
+      // -> complete
+      redraw: this.redraw,
+    };
```

Comment:
The way that this works is each function does something (for example, checking for changes or drawing the graph) and sets `this.state`. If `this.state` is `"complete"`, then `draw()` returns. If `this.state` is something else, the correct function is found in this dictionary and then called.

Even though this is pretty much equivalent to calling the functions directly from other functions, I find it a little easier to follow. That could just be me, though, and I'm interested in hearing other opinions about this kind of design pattern.

Reply:
I think it's easier to follow than having the individual functions call each other. The only think I don't really like is that the return values keep the logic about what state leads to what from being fully encapsulated in the draw function.  It would make more sense to me if 

checkChanges returned something like 'changescomplete' or 'redraw
checkLinks returned something like 'linkscomplete' or 'redraw'
checkNodes returned something like 'nodescomplete' , 'nodestate' or 'redraw'

Reply:
That's a really good suggestion. More intuitive names would make the code a lot easier to follow. What do you think about these state names?

```diff
     const transitions = {
-      // Checks for surface-level changes (e.g. number of nodes)
-      // -> links  : if there are no changes
-      // -> redraw : if there are changes
-      initial: this.checkChanges,
+      // Checks for changes in the number of nodes and links
+      // -> sameSize : if number of nodes and number of  links haven't changed
+      // -> redraw   : if there are changes or if there is no previous graph
+      initial: this.checkGraphSize,

       // Checks for any changes between `links` and `prevLinks`
-      // -> nodes  : if there are no changes
-      // -> redraw : if there are changes
-      links: this.checkLinks,
+      // -> linksIdentical : if there are no changes
+      // -> redraw         : if there are changes
+      sameSize: this.checkLinks,

       // Checks for any changes between `nodes` and `prevNodes`
       // Also checks whether the active node has changed
       // (Sets `this.active` and `this.previousActive`)
-      // -> complete : if there are no changes
-      // -> active   : if the only change is the active node
-      // -> redraw   : if there are any other changes
-      nodes: this.checkNodes,
+      // -> complete      : if there are no changes
+      // -> activeChanged : if the only change is the active node
+      // -> redraw        : if there are any other changes
+      linksIdentical: this.checkNodes,

       // Adds the `active` class to the active node in the graph
       // Removes the `active` class from the previously active node
       // Makes no other change to the graph
       // -> complete
-      active: this.drawActive,
+      activeChanged: this.drawActive,

       // Draws the graph
       // If the graph has already been drawn, this function redraws it
      redraw: this.redraw,
    };
```

Reply:
I like those names. They make the code very clear.